jgrapht-core/src/main/java/org/jgrapht/alg/lca/BinaryLiftingLCAFinder.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * (C) Copyright 2018-2018, by Alexandru Valeanu and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.lca;
+
+import org.jgrapht.Graph;
+import org.jgrapht.Graphs;
+import org.jgrapht.alg.interfaces.LowestCommonAncestorAlgorithm;
+import org.jgrapht.util.VertexToIntegerMapping;
+
+import java.util.*;
+
+import static org.jgrapht.util.MathUtil.log2;
+
+/**
+ * Algorithm for computing lowest common ancestors in rooted trees and forests using the binary lifting method.
+ *
+ * <p>
+ * The method appears in <i>Bender, Michael A., and Martın Farach-Colton. "The level ancestor problem
+ * simplified." Theoretical Computer Science 321.1 (2004): 5-12</i> and it is also nicely presented in
+ * the following article on
+ * <a href="https://www.topcoder.com/community/data-science/data-science-tutorials/range-minimum-query-and-lowest-common-ancestor/#Another%20easy%20solution%20in%20O(N%20logN,%20O(logN)">Topcoder</a>
+ * for more details about the algorithm.
+ * </p>
+ *
+ * <p>
+ * Algorithm idea:<br>
+ * We improve on the naive approach by using jump pointers. These are pointers at a node which reference one of the
+ * node’s ancestors. Each node stores jump pointers to ancestors at levels 1, 2, 4, . . . , 2^k. <br>
+ * Queries are answered by repeatedly jumping from node to node, each time jumping more than half of the
+ * remaining levels between the current ancestor and the goal ancestor (i.e. the lca).
+ * The worst-case number of jumps is $O(log(|V|))$.
+ * </p>
+ *
+ * <p>
+ *  Preprocessing Time complexity: $O(|V| log(|V|))$<br>
+ *  Preprocessing Space complexity:  $O(|V| log(|V|))$<br>
+ *  Query Time complexity: $O(log(|V|))$<br>
+ *  Query Space complexity: $O(1)$<br>
+ * </p>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Alexandru Valeanu
+ */
+public class BinaryLiftingLCAFinder<V, E> implements LowestCommonAncestorAlgorithm<V> {
+
+    private final Graph<V, E> graph;
+    private final Set<V> roots;
+    private final int maxLevel;
+
+    private Map<V, Integer> vertexMap;
+    private List<V> indexList;
+
+    // ancestors[u][i] = the 2^i ancestor of u (e.g ancestors[u][0] = father(u))
+    private int[][] ancestors;
+
+    private int[] timeIn, timeOut;
+    private int clock = 0;
+
+    private int numberComponent;
+    private int[] component;
+
+    /**
+     * Construct a new instance of the algorithm..<br>
```

Comment:
Why two dots?  Also, use p instead of br.

Reply:
I have no idea where the two dots come from. Will change to <p>